docs/design/data-protection-test.md

```diff
@@ -0,0 +1,243 @@
+# DataProtectionTest CRD and Controller Design
+
+## Abstract
+
+This design introduces the **DataProtectionTest** (DPT) CRD and its controller to evaluate data protection performance—including backup upload and volume snapshot readiness—in an OpenShift cluster.. It supports two tests:
+
+1. **Upload Speed Test:** Measures the speed at which a dummy file is uploaded to cloud object storage using configuration from a BackupStorageLocation.
+2. **CSI VolumeSnapshot Test:** Creates a VolumeSnapshot from a specified PVC and measures the time taken (as a duration) for the snapshot to become ready.
+
+Additionally, the design includes a mechanism to determine the S3-compatible vendor.
+
+## Background
+
+In several customer environments, backup/restore operations have experienced delays or stalls due to poor network connectivity, throttling at the cloud provider, or misconfigurations in BackupStorageLocation settings. Given the critical role of backup performance in disaster recovery, administrators need real-time, accurate metrics on how fast data uploads occur and how promptly snapshots are created.
+
+This design addresses that need by measuring:
+
+- **Upload speed:** How quickly data is transferred to object storage.
+- **Snapshot performance:** How long a CSI VolumeSnapshot takes to become ReadyToUse.
+- **S3 compatibility:** Identifying the specific S3 vendor by inspecting HTTP response headers from the storage endpoint.
+
+## Goals
+
+- Upload a test file and compute the speed in Mbps.
+- Create a CSI VolumeSnapshot and measure the time taken for it to become ready (reported as a duration).
+- Identify the S3-compatible vendor using an HTTP call that inspects response headers.
+- Expose all results—including upload speed, snapshot ready duration, and S3 vendor—in the CRD status.
+- Leverage existing BackupStorageLocation (from Velero/DPA) for configuration.
+
+
+## Non-Goals
+
+- This design will not create or modify BackupStorageLocation entries in OADP.
+- It will not implement download or latency tests, focusing solely on upload speed.
+- Scheduling of recurring tests is not supported in the initial version.
+
+## High-Level Design
+Components involved and their responsibilities:
+
+- **DataProtectionTest (DPT) CRD:**
+    - **Spec:**
+        - **backupLocation:** Contains Velero-based backup storage configuration.
+        - **uploadSpeedTestConfig:** Test parameters for the upload (file size, test timeout).
+        - **CSIVolumeSnapshotTestConfig:** Test parameters for the CSI VolumeSnapshot test (snapshot class, source PVC name, and importantly the PVC namespace, plus snapshot timeout).
```

Comment:
why not allow for auto create pvc of specified size to snapshot?
it is important to know how long it takes to provision pvc also for restore perf

Reply:
Discussed this during the sync all. The above was the general consensus.